modin/engines/base/frame/data.py

```diff
@@ -1844,8 +1833,28 @@ def _binary_op(self, op, right_frame, join_type="outer"):
         BasePandasFrame
             A new dataframe.
         """
-        left_parts, right_parts, joined_index = self._copartition(
-            0, right_frame, join_type, sort=True
+        joined_index, ilidx, iridx = self.axes[0].join(
+            right_frame.axes[0], how=join_type, sort=True, return_indexers=True
+        )
```

Comment:
Checking indices for equality before joining will save us a decent time:
```python
import pandas
import numpy as np
from timeit import default_timer as timer


N = 50_000_000

ind1 = pandas.Index(np.arange(N))
ind2 = pandas.Index(np.arange(N))

t1 = timer()
if not ind1.equals(ind2):
    result = ind1.join(ind2, how="outer")
else:
    result = ind1
print("Checking for equality:", timer() - t1)  # ~ 0.1s

t1 = timer()
result = ind1.join(ind2, how="outer")
print("Do not checking for equality:", timer() - t1)  # ~ 8.3s
```
Please also add these check inside `_join_index_objects`

Reply:
@dchigarev good catch!

But I would rather take it out in a separate issue and PR.

Can you create a issue?

Reply:
Done #2490